File: ts/src/llml.ts
import { swagXML } from "./formatters";
import type { Formatters } from "./types";

/**
 * Core LLML function - simple formatter-based API
 */
export const llml = (data: unknown, formatters?: Formatters): string => {
	// Handle no arguments or undefined data
	if (data === undefined) {
		return "";
	}

	// Use default SwagXML formatters if none provided
	const activeFormatters = formatters || swagXML();

	// Iterate through formatters in insertion order
	for (const [predicate, formatFunction] of activeFormatters) {
		if (predicate(data)) {
			return formatFunction(data, llml, activeFormatters);
		}
	}

	// No formatter found - fallback to string conversion
	return String(data);
};

File: py/src/zenbase_llml/llml.py
import typing as t

from .formatters.vibe_xml import Formatters, vibe_xml

__all__ = ["llml", "vibe_xml", "Formatters"]


# Sentinel value to detect when no arguments are provided
_SENTINEL = object()


def llml(data: t.Any = _SENTINEL, formatters: Formatters = vibe_xml) -> str:
    """
    Core LLML function - ultra-simple formatter Map approach
    Takes data and a formatter Map, applies the first matching formatter
    """
    # Handle no arguments case
    if data is _SENTINEL:
        return ""

    # Iterate through formatters in insertion order
    for predicate, format_function in formatters.items():
        if predicate(data):
            return format_function(data, llml, formatters)

    # No formatter found - fallback to string conversion
    return str(data)
